certbot/certbot/_internal/main.py

```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+
+    """
+
+    if config.domains:
+        raise errors.ConfigurationError("You have specified domains, but this function cannot "
+            "be used to modify the domains in a certificate. If you would like to do so, you "
+            "will need to renew the certificate instead of reconfiguring, which can be done "
+            "with the flag --force-renewal. Otherwise, remove the domains from the command "
+            "to continue reconfiguring. You can specify which certificate you want on the command "
+            "line with flag --cert-name instead.")
+    # While we could technically allow domains to be used to specify the certificate in addition to
+    # --cert-name, there's enough complexity with matching certs to domains that it's not worth it,
+    # to say nothing of the difficulty in explaining what exactly this subcommand can modify
+
+
+    # To make sure that the requested changes work, do a dry run. While setting up the dry run,
+    # we will set all the needed fields in config, which will then be saved upon success.
+    config.dry_run = True
+
+    if not config.certname:
+        certname_question = "Which certificate would you like to reconfigure?"
+        config.certname = cert_manager.get_certnames(
+            config, "reconfigure", allow_multiple=False,
+            custom_prompt=certname_question)[0]
+
+    certname = config.certname
+
+    try:
+        renewal_file = storage.renewal_file_for_certname(config, certname)
+    except errors.CertStorageError:
+        raise errors.ConfigurationError(f"An existing certificate with name {certname} could not "
+            "be found. Run `certbot certificates` to list available certificates.")
+
+    lineage_config = copy.deepcopy(config)
+    try:
+        renewal_candidate = renewal.reconstitute(lineage_config, renewal_file)
+    except Exception as e:  # pylint: disable=broad-except
+        raise errors.ConfigurationError(f"Renewal configuration file {renewal_file} "
+            f"(cert: {certname}) produced an unexpected error: {e}.")
+    if not renewal_candidate:
+        raise errors.ConfigurationError("Could not load certificate. See logs for errors.")
+
+    # this is where lineage_config gets fully filled out (e.g. --apache will set auth and installer)
+    installer, auth = plug_sel.choose_configurator_plugins(lineage_config, plugins, "certonly")
+    le_client = _init_le_client(lineage_config, auth, installer)
+
+    # renews cert as dry run to test that the new values are ok
+    # at this point, renewal_candidate.configuration has the old values, but will use
+    # the values from lineage_config when doing the dry run
+    _get_and_save_cert(le_client, lineage_config, certname=certname,
+        lineage=renewal_candidate)
+
+    # this function will update lineage.configuration with the new values, and save it to disk
+    renewal_candidate.save_new_config_values(lineage_config)
```

Comment:
Tracking what was actually updated as opposed to what was there previously...not too bad code-wise, just violates a couple layers of interfaces, which isn't ideal. I can take a stab at it to see how bad it is. Definitely pro printing a success/failure message though, great call.

Reply:
So turns out this is reasonable, as long as we're ok with printing the internal names. Getting the argparse flags back out from the destination is tricky, especially because multiple flags can map onto one destination. I can look into that, but how firmly are you against just printing the internal name? For example, `--nginx` will tell you that `authenticator` and `installer` were updated, but like, that's probably understandable.

Reply:
Currently looks like:

```
$ certbot reconfigure --cert-name example.com --apache --pre-hook "echo new pre hook"
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com

 Successfully updated configuration.
 The following options were added:
    pre_hook: echo new pre hook
 The following options were changed:
    installer: nginx --> apache
    authenticator: nginx --> apache

$ certbot reconfigure --cert-name example.com --apache --pre-hook "echo new pre hook"
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com

 No changes were made to the renewal configuration.
```

Reply:
I think if a straightforward reverse mapping isn't doable, it's better to nix this. Exposing the internal flags in this manner seems antithetical to all the work we did on the UX previously.

Otherwise, I think the output looks great here.

Reply:
Even if there is one for the direct command line options, which I don't think there is, we'd still have to manually add stuff for configurators and such; we're manually doing things like separating `--nginx` into `-i nginx -a nginx`. I only looked briefly though, so I might have missed it in the argparse docs. But then you can still use either `deploy-hook` or `renew-hook` to set `deploy_hook` I believe, so we'd have to go in and manually specify there as well.

The only other thought I have there is just printing the values and not the keys, but that might make things more confusing.

Reply:
OK, let's cut it.

Reply:
Cut. New output:

```
$ certbot reconfigure --cert-name example.com --nginx
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com
Hook 'post-hook' ran with output:
 newpost

No changes were made to the renewal configuration.
$ # stop nginx
$ certbot reconfigure --cert-name example.com --apache
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com
Hook 'post-hook' ran with output:
 newpost

Successfully updated configuration.
Changes will apply when the certificate renews.
```